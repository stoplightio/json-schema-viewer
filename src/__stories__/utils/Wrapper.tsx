import * as React from 'react';

export const Wrapper: React.FunctionComponent<React.HTMLAttributes<HTMLElement>> = ({ children, ...props }) => {
  return (
    <div style={{ padding: '200px auto' }}>
      <div
        // @ts-ignore
        style={{
          width: '100%',
          position: 'relative',
          height: '600px',
          padding: 20,
        }}
        {...props}
      >
        {children}
      </div>
    </div>
  );
};
