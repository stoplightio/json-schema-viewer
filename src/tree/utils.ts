import { isPlainObject } from '@stoplight/json';
import {
  type MirroredSchemaNode,
  type ReferenceNode,
  type RegularNode,
  type SchemaNode,
  isBooleanishNode,
  isReferenceNode,
  isRegularNode,
  isRootNode,
  SchemaNodeKind,
} from '@stoplight/json-schema-tree';
import { BooleanishNode } from '@stoplight/json-schema-tree/nodes/BooleanishNode';

import { isNonNullable } from '../guards/isNonNullable';
import type { ViewMode } from '../types';
import type {
  ArrayNode,
  ComplexArrayNode,
  ComplexDictionaryNode,
  DictionaryNode,
  FlattenableNode,
  PrimitiveArrayNode,
  PrimitiveDictionaryNode,
} from './types';

export type ChildNode = RegularNode | ReferenceNode | MirroredSchemaNode;

export const isNonEmptyParentNode = (
  node: SchemaNode,
): node is RegularNode & {
  children: ChildNode[] & { 0: ChildNode };
} => isRegularNode(node) && !!node.children && node.children.length > 0;

export function isFlattenableNode(node: SchemaNode): node is FlattenableNode {
  if (!isRegularNode(node)) return false;

  if ((!isArrayNode(node) && !isDictionaryNode(node)) || !isNonNullable(node.children) || node.children.length === 0) {
    return false;
  }

  return (
    node.children.length === 1 &&
    ((isRegularNode(node.children[0]) && (!isArrayNode(node) || !isDictionaryNode(node.children[0]))) ||
      (isReferenceNode(node.children[0]) && node.children[0].error !== null))
  );
}

export function isPrimitiveArray(node: SchemaNode): node is PrimitiveArrayNode {
  return isFlattenableNode(node) && isArrayNode(node) && isRegularNode(node.children[0]) && node.children[0].simple;
}

export function isPrimitiveDictionary(node: SchemaNode): node is PrimitiveDictionaryNode {
  return (
    isFlattenableNode(node) && isDictionaryNode(node) && isRegularNode(node.children[0]) && node.children[0].simple
  );
}

export function isComplexArray(node: SchemaNode): node is ComplexArrayNode {
  return isFlattenableNode(node) && isArrayNode(node) && isRegularNode(node.children[0]) && !node.children[0].simple;
}

export function isComplexDictionary(node: SchemaNode): node is ComplexDictionaryNode {
  return (
    isFlattenableNode(node) && isDictionaryNode(node) && isRegularNode(node.children[0]) && !node.children[0].simple
  );
}

export function isDictionaryNode(node: SchemaNode): node is DictionaryNode {
  return (
    isRegularNode(node) &&
    node.primaryType === SchemaNodeKind.Object &&
    isPlainObject(node.fragment.additionalProperties)
  );
}

export function isArrayNode(node: SchemaNode): node is ArrayNode {
  return isRegularNode(node) && node.primaryType === SchemaNodeKind.Array;
}

/**
 * Returns the children of `node` that should be displayed in a viewer or
 * editor. Defaults to `node.children`, except for Arrays that get special
 * handling (flattening).
 */
export function visibleChildren(node: SchemaNode): SchemaNode[] {
  if (!isRegularNode(node) || isPrimitiveArray(node) || isPrimitiveDictionary(node)) {
    return [];
  }
  if (isComplexArray(node) || isComplexDictionary(node)) {
    // flatten the tree here, and show the properties of the item type directly
    return node.children[0].children ?? [];
  }
  return node.children ?? [];
}

export function isPropertyRequired(schemaNode: SchemaNode): boolean {
  const { parent } = schemaNode;
  if (parent === null || !isRegularNode(parent) || schemaNode.subpath.length === 0) {
    return false;
  }

  return !!parent.required?.includes(schemaNode.subpath[schemaNode.subpath.length - 1]);
}

function isValidViewMode(node: RegularNode, viewMode: ViewMode): boolean {
  const { validations } = node;

  if (!!validations.writeOnly === !!validations.readOnly) {
    return true;
  }

  return !((viewMode === 'read' && !!validations.writeOnly) || (viewMode === 'write' && !!validations.readOnly));
}

function isRenderableNode(node: BooleanishNode | RegularNode): boolean {
  if (node.parent === null) return true;

  if (isDictionaryNode(node.parent)) {
    // if dictionary, do not render explicitly defined properties
    return node.subpath.length !== 2 || node.subpath[0] !== 'properties';
  }

  // do not render additionalItems
  if (isArrayNode(node.parent)) {
    return node.subpath[0] !== 'additionalItems';
  }

  // do not render true/false additionalProperties
  if (isRegularNode(node.parent) && node.parent.primaryType === SchemaNodeKind.Object && isBooleanishNode(node)) {
    return !(node.subpath.length === 1 || node.subpath[0] === 'additionalProperties');
  }

  return true;
}

export function shouldNodeBeIncluded(node: SchemaNode, viewMode: ViewMode = 'standalone'): boolean {
  return (
    (isReferenceNode(node) || isRootNode(node) || isRenderableNode(node)) &&
    (!isRegularNode(node) || isValidViewMode(node, viewMode))
  );
}
