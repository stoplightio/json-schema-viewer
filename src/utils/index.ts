export * from './extractVendorExtensions';
export * from './printName';
