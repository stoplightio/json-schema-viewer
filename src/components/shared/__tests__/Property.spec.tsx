import '@testing-library/jest-dom';

import { Provider as MosaicProvider } from '@stoplight/mosaic';
import { render } from '@testing-library/react';
import type { JSONSchema4, JSONSchema7 } from 'json-schema';
import * as React from 'react';
import { describe, expect, it } from 'vitest';

import { SchemaRow } from '../..';
import { buildTree, findNodeWithPath } from './utils';

describe('Property component', () => {
  function renderSchema(schema: JSONSchema4 | JSONSchema7, nodePath?: readonly string[]) {
    const tree = buildTree(schema);

    const node = nodePath ? findNodeWithPath(tree, nodePath) : tree.children[0];

    if (node === undefined) {
      throw new Error('Schema node not found in tree');
    }
    const wrapper = render(
      <MosaicProvider>
        <SchemaRow schemaNode={node} nestingLevel={0} />
      </MosaicProvider>,
    );

    return wrapper;
  }

  it('should render Types with proper type and subtype', () => {
    let schema: JSONSchema4 = {
      type: 'array',
      items: {
        type: 'string',
      },
    };

    let wrapper = renderSchema(schema).queryAllByTestId('property-type')[0];
    expect(wrapper).toMatchInlineSnapshot(
      `
      <span
        class="sl-truncate sl-text-muted"
        data-test="property-type"
      >
        array[string]
      </span>
    `,
    );

    schema = {
      type: 'object',
      additionalProperties: {
        type: 'string',
      },
    };

    wrapper = renderSchema(schema).queryAllByTestId('property-type')[0];
    expect(wrapper).toMatchInlineSnapshot(
      `
      <span
        class="sl-truncate sl-text-muted"
        data-test="property-type"
      >
        array[string]
      </span>
    `,
    );
  });

  it('should handle nullish items', () => {
    const schema: JSONSchema4 = {
      type: 'array',
      items: null as any,
    };

    const wrapper = renderSchema(schema).queryByTestId('property-type');
    expect(wrapper).toMatchInlineSnapshot(
      `
      <span
        class="sl-truncate sl-text-muted"
        data-test="property-type"
      >
        array
      </span>
    `,
    );
  });

  it('should handle nullish $ref', () => {
    const schema: JSONSchema4 = {
      $ref: null as any,
    };

    const wrapper = renderSchema(schema).queryByTestId('property-type-ref');
    expect(wrapper).toMatchInlineSnapshot(
      `
      <span
        class="sl-truncate"
        data-test="property-type-ref"
      >
        $ref
      </span>
    `,
    );
  });

  it('should display true schemas', () => {
    const schema: JSONSchema7 = {
      type: 'object',
      properties: {
        foo: true,
      },
    };

    const wrapper = renderSchema(schema, ['properties', 'foo']).queryByTestId('property-type');
    expect(wrapper).toMatchInlineSnapshot(
      `
      <span
        class="sl-truncate sl-text-muted"
        data-test="property-type"
      >
        any
      </span>
    `,
    );
  });

  it('should display false schemas', () => {
    const schema: JSONSchema7 = {
      type: 'object',
      properties: {
        foo: false,
      },
    };

    const wrapper = renderSchema(schema, ['properties', 'foo']).queryByTestId('property-type');
    expect(wrapper).toMatchInlineSnapshot(
      `
      <span
        class="sl-truncate sl-text-muted"
        data-test="property-type"
      >
        never
      </span>
    `,
    );
  });

  describe('properties names', () => {
    it('given an object, should display the names of its properties', () => {
      const schema: JSONSchema4 = {
        properties: {
          foo: {
            type: 'string',
          },
        },
      };

      const wrapper = renderSchema(schema, ['properties', 'foo']).queryByTestId('schema-row');
      expect(wrapper).toMatchInlineSnapshot(
        `
        <div
          class="sl-flex sl-relative sl-max-w-full sl-py-2"
          data-id="862ab7c3a6663"
          data-test="schema-row"
        >
          <div
            class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
          >
            <div
              class="sl-flex sl-items-center sl-max-w-full"
            >
              <div
                class="sl-flex sl-items-baseline sl-text-base"
              >
                <div
                  class="sl-font-mono sl-font-semibold sl-mr-2"
                  data-test="property-name-foo"
                >
                  foo
                </div>
                <span
                  class="sl-truncate sl-text-muted"
                  data-test="property-type"
                >
                  string
                </span>
              </div>
            </div>
          </div>
        </div>
      `,
      );
    });

    it('given an object among other types, should still display its properties', () => {
      const schema: JSONSchema4 = {
        type: ['string', 'object'],
        properties: {
          foo: {
            type: 'array',
            items: {
              type: 'integer',
            },
          },
        },
      };

      const wrapper = renderSchema(schema, ['properties', 'foo']).queryByTestId('schema-row');
      expect(wrapper).toMatchInlineSnapshot(
        `
        <div
          class="sl-flex sl-relative sl-max-w-full sl-py-2"
          data-id="862ab7c3a6663"
          data-test="schema-row"
        >
          <div
            class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
          >
            <div
              class="sl-flex sl-items-center sl-max-w-full"
            >
              <div
                class="sl-flex sl-items-baseline sl-text-base"
              >
                <div
                  class="sl-font-mono sl-font-semibold sl-mr-2"
                  data-test="property-name-foo"
                >
                  foo
                </div>
                <span
                  class="sl-truncate sl-text-muted"
                  data-test="property-type"
                >
                  array[integer]
                </span>
              </div>
            </div>
          </div>
        </div>
      `,
      );
    });

    it('given an array of objects, should display names of those properties', () => {
      const schema: JSONSchema4 = {
        type: 'array',
        items: {
          properties: {
            foo: {
              type: 'string',
            },
          },
        },
      };

      const wrapper = renderSchema(schema, ['items', 'properties', 'foo']).container.firstChild;
      expect(wrapper).toMatchInlineSnapshot(
        `
        <div
          class=""
          id="mosaic-provider-react-aria-0-1"
        >
          <div
            class=""
            data-overlay-container="true"
          >
            <div
              class="sl-flex sl-relative sl-max-w-full sl-py-2"
              data-id="862ab7c3a6663"
              data-test="schema-row"
            >
              <div
                class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
              >
                <div
                  class="sl-flex sl-items-center sl-max-w-full"
                >
                  <div
                    class="sl-flex sl-items-baseline sl-text-base"
                  >
                    <div
                      class="sl-font-mono sl-font-semibold sl-mr-2"
                      data-test="property-name-foo"
                    >
                      foo
                    </div>
                    <span
                      class="sl-truncate sl-text-muted"
                      data-test="property-type"
                    >
                      string
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      `,
      );
    });

    it('given an array with a combiner inside, should merge it', () => {
      const schema: JSONSchema4 = {
        type: 'array',
        items: {
          oneOf: [
            {
              type: 'string',
            },
            {
              type: 'number',
            },
          ],
        },
      };

      const wrapper = renderSchema(schema).container.firstChild;
      expect(wrapper).toMatchInlineSnapshot(
        `
        <div
          class=""
          id="mosaic-provider-react-aria-0-1"
        >
          <div
            class=""
            data-overlay-container="true"
          >
            <div
              class="sl-flex sl-relative sl-max-w-full sl-py-2"
              data-id="bf8b96e78f11d"
              data-test="schema-row"
            >
              <div
                class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
              >
                <div
                  class="sl-flex sl-items-center sl-max-w-full"
                >
                  <div
                    class="sl-flex sl-items-baseline sl-text-base"
                  >
                    <div
                      class="sl-select sl-relative"
                    >
                      <div
                        aria-hidden="true"
                        style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
                      >
                        <input
                          style="font-size: 16px;"
                          tabindex="-1"
                          type="text"
                        />
                        <label>
                          <select
                            size="2"
                            tabindex="-1"
                          >
                            <option />
                            <option
                              value="0"
                            >
                              array (oneOf) [strings]
                            </option>
                            <option
                              value="1"
                            >
                              array (oneOf) [numbers]
                            </option>
                          </select>
                        </label>
                      </div>
                      <div
                        class="sl-relative"
                      >
                        <button
                          aria-expanded="false"
                          aria-haspopup="listbox"
                          aria-label="Pick a type"
                          aria-labelledby="react-aria-0-4 react-aria-0-6"
                          class="sl-button sl-form-group-border sl-w-full sl-h-sm sl-text-base sl-font-normal sl-px-1.5 sl-bg-transparent sl-rounded sl-border-transparent hover:sl-border-input focus:sl-border-primary active:sl-border-primary sl-border disabled:sl-opacity-60"
                          id="react-aria-0-4"
                          type="button"
                        >
                          <div
                            class="sl-flex sl-flex-1 sl-justify-items-start sl-items-center"
                          >
                            <div
                              class="sl-pr-1"
                            >
                              array (oneOf) [strings]
                            </div>
                          </div>
                          <div
                            class="sl-text-xs sl--mr-0.5 sl-ml-1"
                          >
                            <div
                              class="sl-pt-0.5 sl-pr-0.5"
                            >
                              <svg
                                aria-hidden="true"
                                class="svg-inline--fa fa-chevron-down fa-xs sl-icon"
                                data-icon="chevron-down"
                                data-prefix="fas"
                                focusable="false"
                                role="img"
                                viewBox="0 0 448 512"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M224 416c-8.188 0-16.38-3.125-22.62-9.375l-192-192c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0L224 338.8l169.4-169.4c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25l-192 192C240.4 412.9 232.2 416 224 416z"
                                  fill="currentColor"
                                />
                              </svg>
                            </div>
                          </div>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      `,
      );
    });

    it('given an array with a mergeable combiner inside, should merge it', () => {
      const schema: JSONSchema4 = {
        type: 'array',
        items: {
          oneOf: [
            {
              properties: {
                foo: {},
                bar: {},
                baz: {},
              },
            },
          ],
          type: 'object',
        },
      };

      const wrapper = renderSchema(schema).container.firstChild;
      expect(wrapper).toMatchInlineSnapshot(
        `
        <div
          class=""
          id="mosaic-provider-react-aria-0-1"
        >
          <div
            class=""
            data-overlay-container="true"
          >
            <div
              class="sl-flex sl-relative sl-max-w-full sl-py-2"
              data-id="bf8b96e78f11d"
              data-test="schema-row"
            >
              <div
                class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
              >
                <div
                  class="sl-flex sl-items-center sl-max-w-full sl-cursor-pointer"
                >
                  <div
                    class="sl-flex sl-justify-center sl-w-8 sl--ml-8 sl-pl-3 sl-text-muted"
                    role="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-chevron-down fa-fw fa-sm sl-icon"
                      data-icon="chevron-down"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M224 416c-8.188 0-16.38-3.125-22.62-9.375l-192-192c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0L224 338.8l169.4-169.4c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25l-192 192C240.4 412.9 232.2 416 224 416z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                  <div
                    class="sl-flex sl-items-baseline sl-text-base"
                  >
                    <span
                      class="sl-truncate sl-text-muted"
                      data-test="property-type"
                    >
                      array[object]
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="sl-text-sm"
              data-level="0"
            >
              <div
                class="sl-flex sl-relative sl-max-w-full sl-py-2"
                data-id="85705cc624c84"
                data-test="schema-row"
              >
                <div
                  class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
                >
                  <div
                    class="sl-flex sl-items-center sl-max-w-full"
                  >
                    <div
                      class="sl-flex sl-items-baseline sl-text-base"
                    >
                      <div
                        class="sl-font-mono sl-font-semibold sl-mr-2"
                        data-test="property-name-foo"
                      >
                        foo
                      </div>
                      <span
                        class="sl-truncate sl-text-muted"
                        data-test="property-type"
                      >
                        any
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="sl-flex sl-relative sl-max-w-full sl-py-2"
                data-id="8ce05b74b485f"
                data-test="schema-row"
              >
                <div
                  class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
                >
                  <div
                    class="sl-flex sl-items-center sl-max-w-full"
                  >
                    <div
                      class="sl-flex sl-items-baseline sl-text-base"
                    >
                      <div
                        class="sl-font-mono sl-font-semibold sl-mr-2"
                        data-test="property-name-bar"
                      >
                        bar
                      </div>
                      <span
                        class="sl-truncate sl-text-muted"
                        data-test="property-type"
                      >
                        any
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="sl-flex sl-relative sl-max-w-full sl-py-2"
                data-id="8c605b74b3bb7"
                data-test="schema-row"
              >
                <div
                  class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
                >
                  <div
                    class="sl-flex sl-items-center sl-max-w-full"
                  >
                    <div
                      class="sl-flex sl-items-baseline sl-text-base"
                    >
                      <div
                        class="sl-font-mono sl-font-semibold sl-mr-2"
                        data-test="property-name-baz"
                      >
                        baz
                      </div>
                      <span
                        class="sl-truncate sl-text-muted"
                        data-test="property-type"
                      >
                        any
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      `,
      );
    });

    it('given an array with an allOf inside and enabled allOf merging, should display the name of properties', () => {
      const schema: JSONSchema4 = {
        type: 'object',
        properties: {
          'array-all-objects': {
            type: 'array',
            items: {
              allOf: [
                {
                  properties: {
                    foo: {
                      type: 'string',
                    },
                  },
                },
                {
                  properties: {
                    bar: {
                      type: 'string',
                    },
                  },
                },
              ],
              type: 'object',
            },
          },
        },
      };

      let wrapper = renderSchema(schema, ['properties', 'array-all-objects', 'items', 'properties', 'foo']).container
        .firstChild;
      expect(wrapper).toMatchInlineSnapshot(
        `
        <div
          class=""
          id="mosaic-provider-react-aria-0-1"
        >
          <div
            class=""
            data-overlay-container="true"
          >
            <div
              class="sl-flex sl-relative sl-max-w-full sl-py-2"
              data-id="862ab7c3a6663"
              data-test="schema-row"
            >
              <div
                class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
              >
                <div
                  class="sl-flex sl-items-center sl-max-w-full"
                >
                  <div
                    class="sl-flex sl-items-baseline sl-text-base"
                  >
                    <div
                      class="sl-font-mono sl-font-semibold sl-mr-2"
                      data-test="property-name-foo"
                    >
                      foo
                    </div>
                    <span
                      class="sl-truncate sl-text-muted"
                      data-test="property-type"
                    >
                      string
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      `,
      );

      wrapper = renderSchema(schema, ['properties', 'array-all-objects', 'items', 'properties', 'bar']).container
        .firstChild;
      expect(wrapper).toMatchInlineSnapshot(
        `
        <div
          class=""
          id="mosaic-provider-react-aria-0-1"
        >
          <div
            class=""
            data-overlay-container="true"
          >
            <div
              class="sl-flex sl-relative sl-max-w-full sl-py-2"
              data-id="3cbab69efa81f"
              data-test="schema-row"
            >
              <div
                class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
              >
                <div
                  class="sl-flex sl-items-center sl-max-w-full"
                >
                  <div
                    class="sl-flex sl-items-baseline sl-text-base"
                  >
                    <div
                      class="sl-font-mono sl-font-semibold sl-mr-2"
                      data-test="property-name-bar"
                    >
                      bar
                    </div>
                    <span
                      class="sl-truncate sl-text-muted"
                      data-test="property-type"
                    >
                      string
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      `,
      );
    });

    it('given a ref pointing at complex type, should display property name', () => {
      const schema: JSONSchema4 = {
        properties: {
          foo: {
            $ref: '#/properties/bar',
          },
          bar: {
            type: 'object',
          },
        },
      };

      const tree = buildTree(schema);

      const wrapper = render(<SchemaRow schemaNode={findNodeWithPath(tree, ['properties', 'foo'])!} nestingLevel={0} />)
        .container.firstChild;
      expect(wrapper).toMatchInlineSnapshot(
        `
        <div
          class="sl-flex sl-relative sl-max-w-full sl-py-2"
          data-id="862ab7c3a6663"
          data-test="schema-row"
        >
          <div
            class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
          >
            <div
              class="sl-flex sl-items-center sl-max-w-full"
            >
              <div
                class="sl-flex sl-items-baseline sl-text-base"
              >
                <div
                  class="sl-font-mono sl-font-semibold sl-mr-2"
                  data-test="property-name-foo"
                >
                  foo
                </div>
                <span
                  class="sl-truncate sl-text-muted"
                  data-test="property-type"
                >
                  object
                </span>
              </div>
            </div>
          </div>
        </div>
      `,
      );
    });
  });

  describe('properties titles', () => {
    it('given object type, should render title', () => {
      const schema: JSONSchema4 = {
        title: 'User',
        type: 'object',
        properties: {
          name: {
            type: 'string',
          },
        },
      };

      const wrapper = renderSchema(schema).queryAllByTestId('property-type')[0];
      expect(wrapper).toMatchInlineSnapshot(
        `
        <span
          class="sl-truncate sl-text-muted"
          data-test="property-type"
        >
          User
        </span>
      `,
      );
    });

    it('given array type with non-array items, should render title', () => {
      const schema: JSONSchema4 = {
        type: 'array',
        items: {
          title: 'User',
          type: 'object',
          properties: {
            name: {
              type: 'string',
            },
          },
        },
      };

      const wrapper = renderSchema(schema).queryAllByTestId('property-type')[0];
      expect(wrapper).toMatchInlineSnapshot(
        `
        <span
          class="sl-truncate sl-text-muted"
          data-test="property-type"
        >
          array[User]
        </span>
      `,
      );
    });

    it('given array with no items, should render title', () => {
      const schema: JSONSchema4 = {
        type: 'array',
        title: 'User',
      };

      const wrapper = renderSchema(schema).queryByTestId('property-type');
      expect(wrapper).toMatchInlineSnapshot(
        `
        <span
          class="sl-truncate sl-text-muted"
          data-test="property-type"
        >
          User
        </span>
      `,
      );
    });

    it('given array with defined items, should not render title', () => {
      const schema: JSONSchema4 = {
        type: 'array',
        items: [
          {
            title: 'foo',
            type: 'string',
          },
          {
            title: 'bar',
            type: 'number',
          },
        ],
      };

      const wrapper = renderSchema(schema).queryAllByTestId('property-type')[0];
      expect(wrapper).toMatchInlineSnapshot(
        `
        <span
          class="sl-truncate sl-text-muted"
          data-test="property-type"
        >
          array
        </span>
      `,
      );
    });
  });

  describe('format', () => {
    const schema: JSONSchema4 = require('../../../__fixtures__/formats-schema.json');

    it('should render next to a single type', () => {
      const wrapper = renderSchema(schema, ['properties', 'id']).queryByText('number<float>');
      expect(wrapper).toBeInTheDocument();
    });

    it.each`
      property               | text
      ${'count'}             | ${'integer<int32> or null'}
      ${'date-of-birth'}     | ${'number or string<date-time> or array'}
      ${'array-of-integers'} | ${'array[integer<int32>]'}
      ${'map-of-ids'}        | ${'dictionary[string, integer<int32>]'}
      ${'size'}              | ${'number<byte> or string'}
    `('given $property property, should render next to the appropriate type $text', ({ property }) => {
      const wrapper = renderSchema(schema, ['properties', property]).container.firstChild;
      expect(wrapper).toMatchSnapshot();
    });

    it('should render even when the type(s) is/are missing', () => {
      const wrapper = renderSchema(schema, ['properties', 'notype']).queryByText('<date-time>');
      expect(wrapper).toBeInTheDocument();
    });
  });
});
