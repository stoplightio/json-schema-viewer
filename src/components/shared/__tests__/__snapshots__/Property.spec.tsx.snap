// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Property component > format > given 'array-of-integers' property, should render next to the appropriate type 'array[integer<int32>]' 1`] = `
<div
  class=""
  id="mosaic-provider-react-aria-0-1"
>
  <div
    class=""
    data-overlay-container="true"
  >
    <div
      class="sl-flex sl-relative sl-max-w-full sl-py-2"
      data-id="cba968a2420f0"
      data-test="schema-row"
    >
      <div
        class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
      >
        <div
          class="sl-flex sl-items-center sl-max-w-full"
        >
          <div
            class="sl-flex sl-items-baseline sl-text-base"
          >
            <div
              class="sl-font-mono sl-font-semibold sl-mr-2"
              data-test="property-name-array-of-integers"
            >
              array-of-integers
            </div>
            <span
              class="sl-truncate sl-text-muted"
              data-test="property-type"
            >
              array[integer&lt;int32&gt;]
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Property component > format > given 'count' property, should render next to the appropriate type 'integer<int32> or null' 1`] = `
<div
  class=""
  id="mosaic-provider-react-aria-0-1"
>
  <div
    class=""
    data-overlay-container="true"
  >
    <div
      class="sl-flex sl-relative sl-max-w-full sl-py-2"
      data-id="e093535696d4d"
      data-test="schema-row"
    >
      <div
        class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
      >
        <div
          class="sl-flex sl-items-center sl-max-w-full"
        >
          <div
            class="sl-flex sl-items-baseline sl-text-base"
          >
            <div
              class="sl-font-mono sl-font-semibold sl-mr-2"
              data-test="property-name-count"
            >
              count
            </div>
            <div
              class="sl-truncate"
            >
              <span
                class="sl-truncate sl-text-muted"
                data-test="property-type"
              >
                integer&lt;int32&gt;
              </span>
              <span
                class="sl-text-muted"
              >
                 or 
              </span>
              <span
                class="sl-truncate sl-text-muted"
                data-test="property-type"
              >
                null
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Property component > format > given 'date-of-birth' property, should render next to the appropriate type 'number or string<date-time> or array' 1`] = `
<div
  class=""
  id="mosaic-provider-react-aria-0-1"
>
  <div
    class=""
    data-overlay-container="true"
  >
    <div
      class="sl-flex sl-relative sl-max-w-full sl-py-2"
      data-id="0dd6bfbbfb0ed"
      data-test="schema-row"
    >
      <div
        class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
      >
        <div
          class="sl-flex sl-items-center sl-max-w-full"
        >
          <div
            class="sl-flex sl-items-baseline sl-text-base"
          >
            <div
              class="sl-font-mono sl-font-semibold sl-mr-2"
              data-test="property-name-date-of-birth"
            >
              date-of-birth
            </div>
            <div
              class="sl-truncate"
            >
              <span
                class="sl-truncate sl-text-muted"
                data-test="property-type"
              >
                number
              </span>
              <span
                class="sl-text-muted"
              >
                 or 
              </span>
              <span
                class="sl-truncate sl-text-muted"
                data-test="property-type"
              >
                string&lt;date-time&gt;
              </span>
              <span
                class="sl-text-muted"
              >
                 or 
              </span>
              <span
                class="sl-truncate sl-text-muted"
                data-test="property-type"
              >
                array
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Property component > format > given 'map-of-ids' property, should render next to the appropriate type 'dictionary[string, integer<int32>]' 1`] = `
<div
  class=""
  id="mosaic-provider-react-aria-0-1"
>
  <div
    class=""
    data-overlay-container="true"
  >
    <div
      class="sl-flex sl-relative sl-max-w-full sl-py-2"
      data-id="113c1d8192bed"
      data-test="schema-row"
    >
      <div
        class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
      >
        <div
          class="sl-flex sl-items-center sl-max-w-full"
        >
          <div
            class="sl-flex sl-items-baseline sl-text-base"
          >
            <div
              class="sl-font-mono sl-font-semibold sl-mr-2"
              data-test="property-name-map-of-ids"
            >
              map-of-ids
            </div>
            <span
              class="sl-truncate sl-text-muted"
              data-test="property-type"
            >
              dictionary[string, integer&lt;int32&gt;]
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Property component > format > given 'size' property, should render next to the appropriate type 'number<byte> or string' 1`] = `
<div
  class=""
  id="mosaic-provider-react-aria-0-1"
>
  <div
    class=""
    data-overlay-container="true"
  >
    <div
      class="sl-flex sl-relative sl-max-w-full sl-py-2"
      data-id="0b4cedcd441f1"
      data-test="schema-row"
    >
      <div
        class="sl-stack sl-stack--vertical sl-stack--1 sl-flex sl-flex-1 sl-flex-col sl-items-stretch sl-max-w-full"
      >
        <div
          class="sl-flex sl-items-center sl-max-w-full"
        >
          <div
            class="sl-flex sl-items-baseline sl-text-base"
          >
            <div
              class="sl-font-mono sl-font-semibold sl-mr-2"
              data-test="property-name-size"
            >
              size
            </div>
            <div
              class="sl-truncate"
            >
              <span
                class="sl-truncate sl-text-muted"
                data-test="property-type"
              >
                number&lt;byte&gt;
              </span>
              <span
                class="sl-text-muted"
              >
                 or 
              </span>
              <span
                class="sl-truncate sl-text-muted"
                data-test="property-type"
              >
                string
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
